lib/encumber.rb

```diff
@@ -374,42 +377,74 @@ def move_mouse_to_point(x, y)
 
     end
 
-    def simulate_left_click(xpath, flags=[])
+    def location_for_simulate_mouse_event(original_location, current_location, destination_location, i)
+        tmp_x = (1 - i) * @global_x + i * destination_location[0]
+        tmp_y = (1 - i) * @global_y + i * destination_location[1]
+
+        command('simulateMouseMovedOnPoint', tmp_x, tmp_y, [])
+
+        return [tmp_x, tmp_y]
+    end
+
+    def simulate_left_click(xpath, flags=[], move_mouse=true)
       points = points_for_element(xpath)
-      move_mouse_to_point(points[0], points[1])
+      move_mouse_to_point(points[0], points[1]) if move_mouse
 
-      result = command('simulateLeftClick', id_for_element(xpath), flags)
-      raise "View not found: #{xpath} - #{result}" if result["result"] !='OK'
+      result = command('simulateMouseDown', id_for_element(xpath), flags, $CPLeftMouseDown)
+      raise "An error occured: #{result}" if result["result"] !='OK'
+      result = command('simulateMouseUp', id_for_element(xpath), flags, $CPLeftMouseUp)
+      raise "An error occured: #{result}" if result["result"] !='OK'
     end
 
-    def simulate_left_click_on_point(x, y, flags=[])
-      move_mouse_to_point(x, y)
+    def simulate_left_click_on_point(x, y, flags=[], move_mouse=true)
+      move_mouse_to_point(x, y) if move_mouse
 
-      result = command('simulateLeftClickOnPoint', x, y, flags)
+      result = command('simulateMouseDownOnPoint', x, y, flags, $CPLeftMouseDown)
+      raise "An error occured: #{result}" if result["result"] !='OK'
+      result = command('simulateMouseUpOnPoint', x, y, flags, $CPLeftMouseUp)
+      raise "An error occured: #{result}" if result["result"] !='OK'
     end
 
-    def simulate_double_click(xpath, flags=[])
+    def simulate_right_click(xpath, flags=[], move_mouse=true)
       points = points_for_element(xpath)
-      move_mouse_to_point(points[0], points[1])
+      move_mouse_to_point(points[0], points[1]) if move_mouse
 
-      result = command('simulateDoubleClick', id_for_element(xpath), flags)
-      raise "View not found: #{xpath} - #{result}" if result["result"] != 'OK'
+      result = command('simulateMouseDown', id_for_element(xpath), flags, $CPRightMouseDown)
+      raise "An error occured: #{result}" if result["result"] !='OK'
+      result = command('simulateMouseUp', id_for_element(xpath), flags, $CPRightMouseUp)
+      raise "An error occured: #{result}" if result["result"] !='OK'
     end
 
-    def simulate_double_click_on_point(x, y, flags=[])
-      move_mouse_to_point(x, y)
+    def simulate_right_click_on_point(x, y, flags=[], move_mouse=true)
```

Comment:
For instance in `simulate_double_click` we call two times `simulate_left_click` and this method will call `move_mouse_to_point` after the first click but it's useless and it can slow the double click.
And anyway, the method is expected to achieve a click (not to move the mouse), even if the logic has been implemented like that in all methods.